lib/TH/generic/THStorage.c

```diff
@@ -107,9 +107,9 @@ void THStorage_(free)(THStorage *storage)
   if(!storage)
     return;
 
-  if((storage->flag & TH_STORAGE_REFCOUNTED) && (storage->refcount > 0))
+  if((storage->flag & TH_STORAGE_REFCOUNTED) && (THAtomicGet(&storage->refcount) > 0))
   {
```

Comment:
> Rather I'd raise the question of why you are reading the value beforehand at all (checking that storage->refcount is > 0)? Why not just atomic decrement, and if zero, free?

This was what I was trying to say, though I phrased it poorly.

What I was trying to say: the expression on the right hand side of the && is unnecessary (the read/AtomicGet). 
What I wasn't trying to say: the expression on the right hand side doesn't need to be atomic.